weavertest/multi.go

```diff
@@ -86,26 +90,95 @@ func initMultiProcess(ctx context.Context, t testing.TB, config string) weaver.I
+	//   fromWeaveletReader <--| OS |<-- fromWeaveletWriter
+	//     toWeaveletWriter -->|    |--> toWeaveletReader
+	//                         +----+
+	fromWeaveletReader, fromWeaveletWriter, err := os.Pipe()
 	if err != nil {
-		t.Fatal(err)
+		t.Fatal(fmt.Errorf("cannot create fromWeavelet pipe: %v", err))
 	}
-	ctx = context.WithValue(ctx, runtime.BootstrapKey{}, runtime.Bootstrap{
-		ToWeaveletFd: toWeavelet,
-		ToEnvelopeFd: toEnvelope,
-		TestConfig:   config,
-	})
+	toWeaveletReader, toWeaveletWriter, err := os.Pipe()
+	if err != nil {
+		t.Fatal(fmt.Errorf("cannot create toWeavelet pipe: %v", err))
+	}
+
+	weaveletInfo := createWeaveletForMain(dep)
+
+	// Create a connection between the weavelet for the main process and the envelope.
+	econn, err := conn.NewEnvelopeConn(fromWeaveletReader, toWeaveletWriter, b, weaveletInfo)
+	if err != nil {
+		t.Fatal(fmt.Errorf("cannot create envelope conn: %v", err))
+	}
+
+	// Create and run an envelope for the main process.
+	options := envelope.Options{
+		Restart:         envelope.Never,
+		Retry:           retry.DefaultOptions,
+		GetEnvelopeConn: func() *conn.EnvelopeConn { return econn },
+	}
+	e, err := envelope.NewEnvelope(weaveletInfo, dep.App, b, options)
+	if err != nil {
+		t.Fatal(fmt.Errorf("cannot create envelope: %v", err))
+	}
+
+	go func() {
+		if err := econn.Run(); err != nil {
+			t.Error(err)
+		}
+	}()
+	go func() {
+		if err := e.Run(ctx); err != nil {
+			t.Error(err)
+		}
+	}()
 
 	t.Cleanup(func() {
 		cancelFunc()
+
+		// There is a very subtle bug we have to avoid. As explained in the
+		// official godoc [1], if an *os.File is garbage collected, the underlying
+		// file is closed. Thus, if toWeaveletReader or fromWeaveletWriter are
+		// garbage collected, the underlying pipes are closed.
+		//
+		// The main weavelet, which is running in a separate goroutine, is reading
+		// from and writing to these pipes, but it doesn't use toWeaveletReader or
+		// fromWeaveletWriter directly. Instead components to start., it constructs new files using the
+		// two pipe's file descriptors. This means that Go is free to garbage
+		// collect fromWeaveletWriter and toWeaveletWriter even though the
+		// underlying pipes are still being used by the main weavelet.
+		//
+		// If the pipes get closed while the main weavelet is running, then the
+		// weavelet can crash with all sorts of errors (e.g., bad pipe, resource
+		// temporarily unavailable, nil pointer dereferences, etc). You can run [2]
+		// locally to reproduce this behavior. If the program doesn't crash on your
+		// machine, try increasing the number of iterations.
+		//
+		// To avoid this, we have to ensure that the pipes are not garbage
+		// collected prematurely. This is done by holding a reference to the
+		// os.File in the closure passed to t.Cleanup
+		//
+		// [1]: https://pkg.go.dev/os#File.Fd
+		// [2]: https://go.dev/play/p/9JG7voL2oHP
+		_ = fromWeaveletWriter
```

Comment:
Hm, I'm worried it's still possible for a test to end, these variables get garbage collected, the pipes get closed, and a subprocess weavelet with access to these pipes crashes? I think we have to wait for all subprocesses to die (which we don't have a good way of doing), not wait for the test to end?

What do you think about sticking with the super janky package-scoped `dontGC` for now, and then at some point in the future, when we clean up the code to make sure subprocesses are dying when they're supposed to, we can get rid of it?

Reply:
I think the scenario you describe is very likely, but I wonder if it's a problem. As long as the test is finished, and nothing tries to write to `t.Log` after the test has finished (which seems to be handled by `logSaver` in this code), does it matter if a process crashes? I believe any logs and errors would be dropped (not sent to stderr). Having the subprocesses exit earlier seems preferable (even if they are exiting with an error). I ran the tests with `-count=100` a few times to see if I could get any tests to fail, but they passed consistently as they do on `main`.

I'm happy to switch back to the package-scoped variable if you prefer that approach. What do you think?

Reply:
Let's switch back to the package-scoped variable. I think what I wrote above was wrong. The main weavelet is actually running in this process, not a subprocess, and is the thing that can crash.

Reply:
Ok, sounds fair! I've restored the package-scoped `dontGC` slice in the latest commit. I added a package-scoped `dontGCLock sync.Mutex` variable as well, since we don't have access to the mutex that was used previously used. 

The previous lock was effectively scoped to the test I believe. It was a field on `Babysitter`, which is created for each call to `weavertest.Init`. I did not realize this until just now, but I think the previous implementation had a data race if two tests cases in the same package used `t.Parallel` and both of them called `weavertest.Init` at the start of the test. The package-scoped mutex should guard against that case.

Reply:
Welp, yeah the old code was completely busted! I told you the code was janky 🤣. Thanks for finding and fixing.